sherpa/astro/ui/utils.py

```diff
@@ -10433,7 +10417,13 @@ def get_source_component_plot(self, id, model=None, recalc=True):
         if isinstance(model, string_types):
             model = self._eval_model_expression(model)
 
-        d = self.get_data(id)
+        try:
+            d = self.get_data(id)
+        except IdentifierErr as ie:
+            if recalc:
+                raise ie
+            d = None
+
```

Comment:
Since this code turns up three times, should it be a separate function?

Reply:
I think I'd like to check that we're doing something sensible, and worth encapsulating, first!

Maybe, in this case, we should require a dataset to be present, to make sure that you're changing the "correct" plot type (the datatype-specific plots is new functionality added in 4.12.2 or 4.13.0, I forget which)? I don't like that (I want to be able to set prefs before loading data), but requiring a dataset is cleaner.

Or, we can add a dtype parameter (name open for bikeshedding) which defines the plot type to change? Then we need to decide whether it replaces id or only comes into play if there is no dataset available.

Or we can say "when you change the plto preferences, you change all datatype preferences" (modulo some issues about keys not being always present), but that is *hard* to retrofit into the existing system.

Reply:
I apparently did not understand the intention of this change. What is the point of `get_xxx_plot` when there is no data to be plotted? It's to get it of the `get_plot_prefs` functions, right? If so, the code as suggest will do that (as you showed). I know you are working on a bigger development that goes beyond that, so I don't know exactly how that fits in there.

Conceptually, `get_plot_xxx` makes more sense if there is data to plot. So, I guess we should require data to be present, however I'll comment more in #1096.

Reply:
Previously, we had a patch-work of code that didn't make it easy to work out responsibility, and the same bit of logic was scattered about. I am trying to make the `get_xxx_plot` calls be the place that maps from identifier to relevant plot object. These calls have existed for a long time (i.e. they're not new).

The move to plot-specific types has complicated the interface, but it already somewhat existed (handling of PHA data).

Whether this is the correct semantics of the `get_xxx_plot` (and other routines) is open to question.

Reply:
To be clear: We can just accept this and move and possibly rework it in the next PR. This is really just a minor suggest and I don't want to hold up all the rest of the PR for this one thing that can be decided later. It's really only a few lines. So, I'm going to accept this and continue the discussion in #1097.